src/main/java/org/opensearch/neuralsearch/sparse/NeuralSparseIndexShard.java

```diff
@@ -0,0 +1,268 @@
+                segmentInfo.dir,
+                segmentInfo,
+                leafReader.getFieldInfos(),
+                IOContext.DEFAULT
+            );
+
+            // Create SparseTermsLuceneReader for inverted index
+            final SparseTermsLuceneReader luceneReader = new SparseTermsLuceneReader(readState);
+
+            // Create CacheGated readers
+            final CacheGatedForwardIndexReader forwardIndexReader = getCacheGatedForwardIndexReader(binaryDocValues, key, docCount);
+            final CacheGatedPostingsReader postingsReader = new CacheGatedPostingsReader(
+                fieldInfo.name,
+                InMemoryClusteredPosting.getOrCreate(key).getReader(),
+                InMemoryClusteredPosting.getOrCreate(key).getWriter(),
+                luceneReader
+            );
+
+            // Warm up forward index
+            int loadedDocs = 0;
+            for (int docId = 0; docId < docCount; docId++) {
+                if (forwardIndexReader.read(docId) != null) {
+                    loadedDocs++;
+                }
+            }
+
+            // Warm up inverted index
+            final Set<BytesRef> terms = postingsReader.getTerms();
+            for (BytesRef term : terms) {
+                postingsReader.read(term);
+            }
+
+            log.info("[Neural Sparse] Warmed up {} docs and {} terms for field: {}", loadedDocs, terms.size(), fieldInfo.name);
+
+        } catch (Exception e) {
+            log.warn("[Neural Sparse] Failed to create lucene readers, using forward index only", e);
+
+            // Fallback: only warm up forward index
+            final CacheGatedForwardIndexReader forwardIndexReader = getCacheGatedForwardIndexReader(binaryDocValues, key, docCount);
+            int loadedDocs = 0;
+            for (int docId = 0; docId < docCount; docId++) {
+                if (forwardIndexReader.read(docId) != null) {
+                    loadedDocs++;
+                }
+            }
+            log.info("[Neural Sparse] Warmed up {} documents for field: {} (forward index only)", loadedDocs, fieldInfo.name);
+        }
+    }
+
+    /**
+     * Create CacheGatedForwardIndexReader following the existing pattern
+     */
+    private CacheGatedForwardIndexReader getCacheGatedForwardIndexReader(
+        BinaryDocValues binaryDocValues,
+        InMemoryKey.IndexKey key,
+        int docCount
+    ) {
+        if (binaryDocValues instanceof SparseBinaryDocValuesPassThrough sparseBinaryDocValues) {
+            InMemorySparseVectorForwardIndex index = InMemorySparseVectorForwardIndex.get(key);
+            if (index == null) {
+                index = InMemorySparseVectorForwardIndex.getOrCreate(key, docCount);
+            }
+            return new CacheGatedForwardIndexReader(null, index.getWriter(), sparseBinaryDocValues);
+        } else {
+            // For regular BinaryDocValues, create a SparseVectorReader wrapper
+            final SparseVectorReader luceneReader = docId -> {
```

Comment:
If binaryDocValues is not instanceof SparseBinaryDocValuesPassThrough, then just wrap it with SparseBinaryDocValuesPassThrough, although I think this might be an impossible situation